pep-0712.rst

```diff
@@ -0,0 +1,326 @@
+
+Therefore, this PEP adds a ``converter`` parameter to :func:`dataclasses.field`
+(along with the requisite changes to :class:`dataclasses.Field` and
+:func:`~py3.11:typing.dataclass_transform`) to specify the function to use to
+convert the input value for each field to the representation to be stored in
+the dataclass.
+
+Motivation
+==========
+
+There is no existing, standard way for :mod:`dataclasses` or third-party
+dataclass-like libraries to support argument conversion in a type-checkable
+way. To work around this limitation, library authors/users are forced to choose
+to:
+
+* Opt-in to a custom Mypy plugin. These plugins help Mypy understand the
+  conversion semantics, but not other tools.
+* Shift conversion responsibility onto the caller of the dataclass
+  constructor. This can make constructing certain dataclasses unnecessarily
+  verbose and repetitive.
+* Provide a custom ``__init__`` and which declares "wider" parameter types and
+  converts them when setting the appropriate attribute. This not only duplicates
+  the typing annotations between the converter and ``__init__``, but also opts
+  the user out of many of the features :mod:`dataclasses` provides.
+* Not rely on, or ignore type-checking.
+
+None of these choices are ideal.
+
+Rationale
+=========
+
+Adding argument conversion semantics is useful and beneficial enough that most
+dataclass-like libraries provide support. Adding this feature to the standard
+library means more users are able to opt-in to these benefits without requiring
+third-party libraries. Additionally third-party libraries are able to clue
+type-checkers into their own conversion semantics through added support in
+:func:`~py3.11:typing.dataclass_transform`, meaning users of those libraries
+benefit as well.
+
+Specification
+=============
+
+New ``converter`` parameter
+---------------------------
+
+This specification introduces a new parameter named ``converter`` to the
+:func:`dataclasses.field` function. When an ``__init__`` method is synthesized
+by ``dataclass``-like semantics, if an argument is provided for the field, the
+``dataclass`` object's attribute will be assigned the result of calling the
+converter on the provided argument. If no argument is given and the field was
+constructed with a default value, the ``dataclass`` object's attribute will be
+assigned the result of calling the converter on the provided default.
+
+Adding this parameter also implies the following changes:
+
+* A ``converter`` attribute will be added to :class:`dataclasses.Field`.
+* ``converter`` will be added to the field specifier parameters of arguments
+  provided to :func:`~py3.11:typing.dataclass_transform`'s ``field`` parameter.
+
+Example
+'''''''
+
+.. code-block:: python
+
+  @dataclasses.dataclass
+  class InventoryItem:
+      # `converter` as a type
+      id: int = dataclasses.field(converter=int)
+      skus: tuple[int, ...] = dataclasses.field(converter=tuple[int])
```

Comment:
Do you mean `tuple[int, ...]`? This type would mean a one-element tuple.

Reply:
Oops—fixed that one place in that line (and others) but somehow missed the second one. I think we got all the others already.